File: Stack/implement-stack-using-queues-225.py
# https://leetcode.com/problems/implement-stack-using-queues/

# TL;DR:
## This approach uses O(n) for push operation and O(1) for pop
## In push we append the element to queue2, take all elements from queue1 and put them in queue2 and then put all elements of queue2 back to queue1
## This assures the following:
##  - we add the new element as the first one in queue2 and then all of the previous elements from queue1 follow him in queue2
##  - we then move the current queue2 back to queue1 (the primary queue) so that we have the newly added element as the first one in the queu (mimicking LIFO)

# Complexities:
## Time => O(n) for push, O(1) for pop, where n is the amount of numbers passed as input
## Space => O(n), where n is the amount of numbers passed as input

class MyStack:
    def __init__(self):
        self.queue1 = collections.deque()
        self.queue2 = collections.deque()

    def push(self, x: int) -> None:
        self.queue2.append(x)

        while self.queue1:
            self.queue2.append(self.queue1.popleft())

        while self.queue2:
            self.queue1.append(self.queue2.popleft())

    def pop(self) -> int:
        return self.queue1.popleft()

    def top(self) -> int:
        return self.queue1[0]

    def empty(self) -> bool:
        return not self.queue1

File: Stack/implement-stack-using-queues-225.ts
// https://leetcode.com/problems/implement-stack-using-queues/

// Hint:
// - Implement the stack with 2 queues and then optimize to use only 1 queue

// TL;DR:
// Use a queue to implement the stack
// Push:
//   - Push the new element to the queue
//   - For each element in the queue, push it to the back of the queue (except the last one which is the new element)
// Pop:
//   - Pop the top of the queue
// Top:
//   - Return the top of the queue
// Empty:
//   - Return true if the queue is empty, false otherwise

// Complexities:
// Time => O(n) for push, O(1) for pop, O(1) for top, O(1) for empty
// Space => O(n), where n is the number of elements in the stack

class MyStack {
	private readonly mainQ: Queue<number>;

	constructor() {
		this.mainQ = new Queue();
	}

	push(x: number): void {
		this.mainQ.push(x);
		for (let i = 0; i < this.mainQ.size() - 1; i++) {
			this.mainQ.push(this.mainQ.pop()!);
		}
	}

	pop(): number {
		return this.mainQ.pop()!;
	}

	top(): number {
		return this.mainQ.front();
	}

	empty(): boolean {
		return this.mainQ.isEmpty();
	}
}
